sel4-extract/sel4-extract.c

```diff
@@ -0,0 +1,384 @@
+// SPDX-License-Identifier: GPL-2.0-only
+/*
+ * Copyright 2023, Technology Innovation Institute
+ *
+ */
+#include <stdio.h>
+#include <stdint.h>
+#include <sys/mman.h>
+#include <stdlib.h>
+#include <fcntl.h>
+#include <errno.h>
+#include <string.h>
+#include <stdbool.h>
+#include <assert.h>
+
+#define DEBUG 1
+
+#define PAGE_SIZE 4096
+
+#define ARRAY_SIZE(a) (sizeof(a)/sizeof(a[0]))
+#define BIT(x) (1 << (x))
+
+#if DEBUG
+#define pr_dbg(...) do { printf("[DBG]: " __VA_ARGS__); printf("\t[%d@%s]\n", __LINE__, __func__); fflush(stdout); } while (0)
+#else
+#define pr_dbg(...)
+#endif
+
+typedef enum {
+	Entry_Interrupt,
+	Entry_UnknownSyscall,
+	Entry_UserLevelFault,
+	Entry_DebugFault,
+	Entry_VMFault,
+	Entry_Syscall,
+	Entry_UnimplementedDevice,
+	Entry_VCPUFault,
+	Entry_Switch,
+} entry_type_t;
+
+char *entry_names[] = {
+	"Entry_Interrupt",
+	"Entry_UnknownSyscall",
+	"Entry_UserLevelFault",
+	"Entry_DebugFault",
+	"Entry_VMFault",
+	"Entry_Syscall",
+	"Entry_UnimplementedDevice",
+	"Entry_VCPUFault",
+	"Entry_Switch",
+};
+
+#define SEL4_PACKED             __attribute__((packed))
+typedef uint64_t seL4_Word;
+
+/**
+ * @brief Kernel entry logging
+ *
+ * Encapsulates useful info about the cause of the kernel entry
+ */
+typedef struct SEL4_PACKED kernel_entry {
+	seL4_Word path : 4;
+	union {
+		struct {
+			seL4_Word core : 3;
+			seL4_Word word : 26;
+		};
+		/* Tracked kernel entry info filled from outside this file */
+		struct {
+			seL4_Word syscall_no     : 4;
+			seL4_Word cap_type       : 5;
+			seL4_Word is_fastpath    : 1;
+			seL4_Word invocation_tag : 19;
+		};
+		void *next;
+	};
+} kernel_entry_t;
+
+typedef struct benchmark_syscall_log_entry {
+	uint64_t  start_time;
+	uint32_t  duration;
+	kernel_entry_t entry;
+} benchmark_track_kernel_entry_t;
+
+struct Page_Header {
```

Comment:
Do you have strong reason to stay with this Camel_Case capitalization? If not, could you just follow like the one used in Linux kernel, like `struct entry_header` and `struct event_function`.

Reply:
Sorry, I was really thinking that I fixed that already. To many different repos are opened simultaneous on my laptop. Probably it time to keep vim and adopt some proper IDE.